File: CPP/OOP/AutoPen.cpp
#ifndef AUTOPEN_CPP
#define AUTOPEN_CPP

#include "Pen.cpp"

class AutoPen : public Pen {
	private:
		bool open;

	public:
		AutoPen(int amount = 100) : Pen(amount), open(false) {}

		bool isOpen() const { return open; }

		void click() { open = !open; }

		virtual void write(Paper& paper, const string& text) {
			#ifdef DEBUG
				cout << "AutoPen writing..." << endl;
			#endif

			if ( open ) {
				Pen::write(paper, text);
			}
		}
};

#endif

File: OtherLang/AutoPen.java
public class AutoPen extends Pen {
	private boolean isOpen = false;

	public AutoPen(int ink) throws Throwable {
		super(ink);
	}

	public void write(Paper paper, String text) throws Throwable {
		if ( !isOpen ) {
			throw new ClosedPen();
		}
		super.write(paper, text);
	}
}